pkg/datastore/mysql/mysql.go

```diff
@@ -0,0 +1,131 @@
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"io/ioutil"
+
+	// mysql driver
+	_ "github.com/go-sql-driver/mysql"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+	"go.uber.org/zap"
+)
+
+// MySQL client wrapper
+type MySQL struct {
+	client       *sql.DB
+	logger       *zap.Logger
+	usernameFile string
+	passwordFile string
+}
+
+// Option for create MySQL typed instance
+type Option func(*MySQL)
+
+// WithLogger returns logger setup function
+func WithLogger(logger *zap.Logger) Option {
+	return func(m *MySQL) {
+		m.logger = logger
+	}
+}
+
+// WithAuthenticationFile returns auth info setup function
+func WithAuthenticationFile(usernameFile, passwordFile string) Option {
+	return func(m *MySQL) {
+		m.usernameFile = usernameFile
+		m.passwordFile = passwordFile
+	}
+}
+
+// NewMySQL returns new MySQL instance
+func NewMySQL(url, database string, opts ...Option) (*MySQL, error) {
+	m := &MySQL{
+		logger: zap.NewNop(),
+	}
+	for _, opt := range opts {
+		opt(m)
+	}
+
+	dataSourceName, err := m.buildDataSourceName(url, database)
+	if err != nil {
+		return nil, err
+	}
+	db, err := sql.Open("mysql", dataSourceName)
+	if err != nil {
+		return nil, err
+	}
+	m.client = db
+
+	return m, nil
+}
+
+// Find implementation for MySQL
+func (m *MySQL) Find(ctx context.Context, kind string, opts datastore.ListOptions) (datastore.Iterator, error) {
+	return nil, datastore.ErrUnimplemented
+}
+
+// Get implementation for MySQL
+func (m *MySQL) Get(ctx context.Context, kind, id string, v interface{}) error {
+	return datastore.ErrUnimplemented
+}
+
+// Create implementation for MySQL
+func (m *MySQL) Create(ctx context.Context, kind, id string, entity interface{}) error {
+	return datastore.ErrUnimplemented
+}
+
+// Put implementation for MySQL
+func (m *MySQL) Put(ctx context.Context, kind, id string, entity interface{}) error {
+	return datastore.ErrUnimplemented
+}
+
+// Update implementation for MySQL
+func (m *MySQL) Update(ctx context.Context, kind, id string, factory datastore.Factory, updater datastore.Updater) error {
+	return datastore.ErrUnimplemented
+}
+
+// Close implementation for MySQL
+func (m *MySQL) Close() error {
+	return m.client.Close()
+}
+
+func (m *MySQL) buildDataSourceName(url, database string) (string, error) {
+	if url == "" {
+		return "", fmt.Errorf("url is required field")
+	}
+	if database == "" {
+		return "", fmt.Errorf("database is required field")
+	}
+	// In case username and password files are not provided,
+	// those values may included in the URL already so just return the given URL.
+	if m.usernameFile == "" || m.passwordFile == "" {
+		return url, nil
+	}
+
+	username, err := ioutil.ReadFile(m.usernameFile)
+	if err != nil {
+		return "", fmt.Errorf("failed to read username file: %w", err)
+	}
+	password, err := ioutil.ReadFile(m.passwordFile)
+	if err != nil {
+		return "", fmt.Errorf("failed to read password file: %w", err)
+	}
+
+	return fmt.Sprintf("%s:%s@tcp(%s)/%s", username, password, url, database), nil
+}
```

Comment:
Could you add some tests for this function?

Reply:
Sure, addressed on https://github.com/pipe-cd/pipe/pull/1641/commits/5f0dcc9d6221ba6019f02e5ae17b9ed3252c0279 🙏